### PR TITLE
introduce a contract for extending/overriding layout templates

### PR DESCRIPTION
**Pull request by KorzunYuri**
I propose a contract for components that:
- may have their layouts customized in other projects
- have multiple layouts by design (e.g. horizontal/vertical tabs)

## Description
Contract includes:
1. an interface to be implemented by component models
2. a dialog field to be included by components when layout choice is required
3. rules that component should follow:
- implement MultiTemplateComponent interface (no overriding required until we need a choice of layout)
- have default layout in ./templates/template-default.html file

More detailed explanation is in MultiTemplateComponent javadoc.

## Motivation and Context
It's always better to do the same things the same way, but up to this moment we haven't had a contract for component layouts naming and placement, as well as for the way we override/add layouts.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) labeled with `bug`
- [x] New feature (non-breaking change which adds functionality) labeled with `enhancement`
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code standards](../CONTRIBUTING.md) of this project.
- [ ] My change requires updating the documentation. I have updated the documentation accordingly.

### CODE REVIEW

File: applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/MultiTemplateComponent.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  /**
+   * Return layout template relative path. Returns default template path by default,
+   * so there is no need to override the method and add a property to you component
+   * if you will only have one layout - just place the layout in the template-default.html file
+   *
+   * @return layout template's relative path
+   */
+  default String getTemplatePath() {
+    return "./templates/template-default.html";
+  }
```
Comment (stan-r-ds) [inline applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/MultiTemplateComponent.java]: The name of this interface contains **Multi**Template, but the method to be implemented expects just a **single** template. Probably some rewording of those names would give better understandability of their usage 💭

File: applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/package-info.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (C) 2023 Dynamic Solutions
```
Comment (stan-r-ds) [inline applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/package-info.java]: 2023 here, vs 2024 in the interface 💭

File: applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/MultiTemplateComponent.java (hunk 1 above)
Comment (KorzunYuri) [reply applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/MultiTemplateComponent.java]: Component will always have only one template for rendering per time unit. The problem is that my contract covers two topics: 1) extensibility (having multiple templates to choose from); 2) default template naming (always knowing where default template is to be able to rewrite it). So it's not **_always_** MultiTemplate, it is **_potentially_** multi-template. And I'm not sure how I should name this interface. Do you have an idea?

Comment (paweljozwicki) [reply applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/MultiTemplateComponent.java]: What do you think about the TemplateHandler?

Comment (KorzunYuri) [reply applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/MultiTemplateComponent.java]: I like it! If there is no other proposals, I'll use it.

**Review by stan-r-ds — APPROVED** [review]

**Comment by KorzunYuri** [comment]
Interface was renamed to TemplateHandler. Respective changes were made in [Quote component PR](https://github.com/websight-io/kyanite/pull/454)

File: applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/TemplateHandler.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ *    or add a new one</li>
+ *    <li>you are developing a component that can have different layouts by design
+ *    (e.g. horizontal and vertical tabs)</li>
+ * </ul>
+ * </p>
+ *
+ * <p>Component must follow this rules:
+ * <ol>
+ *    <li>component has a default template under templates/template-default.html</li>
+ *    <li>component's model implements MultiTemplateComponent interface</li>
```
Comment (paweljozwicki) [inline applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/TemplateHandler.java]: Let's update the name in the javadoc

Comment (KorzunYuri) [reply applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/layouts/TemplateHandler.java]: I've updated the javadoc and hope the wording is a little clearer now.

**Review by paweljozwicki — APPROVED** [review]

**Pull request merged**